### DISCUSSION TITLE
Example: Applying style and formatting rules to content

### DISCUSSION THREAD

**Discussion by siralmat** [jglev/obsidian-apply-patterns-plugin#55] [Example Patterns]
I use Obsidian to collect quotes and reference materials from around the web, and they often come with formatting quirks - for example, extra whitespace around list items and headers. I use a pattern to clean up notes so they fit my personal preferences:

1. Replace smart apostrophes (`‘’`) with straight apostrophes (`'`)
2. Replace smart quotes (`“”`) with straight quotes (`"`)
3. Strip trailing spaces and tabs from the end of lines
4. Remove extra leading spaces from list items
5. Remove blank lines after headers 
6. Delete invisible "anchor" headers (sometimes you get these when pasting web content; see example below)
7. Replace double spaces after a full stop with a single space
8. Shorten URL display text: e.g. `https://example.com/` and `www.example.com` are shortened to `example.com` (actual URL destinations are not affected)

Here's an example with the Apply Patterns readme. Left is how it appears directly after copying and pasting into Obsidian, right is after applying my pattern:
![image](https://user-images.githubusercontent.com/34409782/180211629-2976a4e3-57e3-494b-8abc-26c37e6cb9e9.png)

The rules are meant to be safe for general use and unlikely to make unexpected or destructive changes to content, but I **don't recommend** running the pattern on code or other structured content where whitespace might be important. 

I'm sure most people don't share my exact formatting tastes, but hopefully these are helpful for anyone trying to do something similar. If anyone has suggestions for improvement or can find any edge cases that are handled incorrectly, I'd love to hear about them. 🙂

```json
  {
    "name": "Cleanup",
    "rules": [
      {
        "from": "[‘’]",
        "to": "'",
        "caseInsensitive": false,
        "global": true,
        "multiline": false,
        "sticky": false,
        "disabled": false
      },
      {
        "from": "[“”]",
        "to": "\"",
        "caseInsensitive": false,
        "global": true,
        "multiline": false,
        "sticky": false,
        "disabled": false
      },
      {
        "from": "[ \\t]+$",
        "to": "",
        "caseInsensitive": false,
        "global": true,
        "multiline": true,
        "sticky": false,
        "disabled": false
      },
      {
        "from": "^(\\s*([-*]|\\d\\.))\\s{2,}",
        "to": "$1 ",
        "caseInsensitive": false,
        "global": true,
        "multiline": true,
        "sticky": false,
        "disabled": false
      },
      {
        "from": "(^#{2,} .+$)(\\n){2,4}",
        "to": "$1\\n",
        "caseInsensitive": false,
        "global": true,
        "multiline": true,
        "sticky": false,
        "disabled": false
      },
      {
        "from": "^#{2,} \\[\\]\\(.+#.+\\)$\\n+",
        "to": "",
        "caseInsensitive": false,
        "global": true,
        "multiline": true,
        "sticky": false,
        "disabled": false
      },
      {
        "from": "\\.  (\\w)",
        "to": ". $1",
        "caseInsensitive": false,
        "global": true,
        "multiline": false,
        "sticky": false,
        "disabled": false
      },
      {
        "from": "\\[(https?://)?(www\\.)?(.+?)\\/?\\]\\(",
        "to": "[$3](",
        "caseInsensitive": false,
        "global": true,
        "multiline": false,
        "sticky": false,
        "disabled": false
      }
    ],
    "collapsed": false,
    "cursorRegexStart": "$",
    "cursorRegexEnd": "^"
  },
```

**Comment by cs-mshah** [comment]
Can you add something to skip the code part? Other than that this is really useful when using it with clipper.

**Reply by siralmat** [reply]
I'm glad you find it useful! 😄 

Do you mean skipping over codeblocks without applying rules to the contents? It's an interesting idea - but it's very hard for a regex pattern to "know" if it's inside a codeblock or not. Here's an example to show what I mean:
<img width="323" alt="image" src="https://user-images.githubusercontent.com/34409782/198952071-c32273ad-bcbc-442f-8a7e-3383fc2e6a6b.png">
You could use [lookarounds](https://www.regular-expressions.info/lookaround.html) to only match the "outside" sections, but it's tricky to do properly and will make the rules very complicated. 

If you really need this functionality, another approach could be to modify the Apply Patterns plugin. You could write a new 'Apply to plaintext in document' option that skips over codeblocks and only runs the pattern on the rest of the note. I think this would be a bit easier to implement and means you don't have to rewrite any of the patterns.

I hope that helps!
